### PR TITLE
FEAT: Cleanup Directory Based on File Type

### PR DESCRIPTION
**Pull request by mubarak-ramadan**
https://github.com/users/yahayaohinoyi/projects/4/views/1?pane=issue&itemId=79800120

- [x] Feature: Clean files based on specific file types or extensions.
- [x] Use Case: Users can target specific files like .log, .tmp, or .bak files for removal.
- [x] Unit tests

### CODE REVIEW

**Comment by yahayaohinoyi** [comment]
could you add the .idea/* folder to gitignore? I think it's a thing with rust rover environment

File: src/arg.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+use clap::{Arg, Command};
```
Comment (yahayaohinoyi) [inline src/arg.rs:1]: Question: Is there ea reason you prefer the clap's arg command to the one from rust's standard library?

**Comment by yahayaohinoyi** [comment]
> could you add the .idea/* folder to gitignore? I think it's a thing with rust rover environment

It's still in the commit 6e7cf87bb9f59ede505bb3542d8302afe4f804e2

**Comment by yahayaohinoyi** [comment]
can you also run` cargo format ` to format the code? Maybe we should consider adding it to our CI at some point? Or a pre-commit hook?

File: src/arg.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                .long("dir")
+                .required(true)
+                .help("Directory to clean up")
+        )
+        .arg(
+            Arg::new("types")
+                .short('t')
+                .long("types")
+                .required(false)
+                .value_delimiter(',')
```
Comment (yahayaohinoyi) [inline src/arg.rs]: kinda curious here, how would the command line args look like?

CMD_EXEC log,bak,others...? should we make them space separated instead?

File: src/arg.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
+use clap::{Arg, Command};
+
+const APP: &str = "Directory cleaner";
+pub struct Args {
+    pub types: Vec<String>,
+    pub min_size: u64,
```
Comment (yahayaohinoyi) [inline src/arg.rs:6]: what does min size do?

File: src/main.rs [hunk 4]
```diff
@@ -1,3 +1,89 @@
...
+
+    println!("Cleaning completed successfully.");
+
+    Ok(())
 }
+
+
+fn directory_cleaner(dir: &String, paths_to_clear: &[String], size: u64) -> Result<()> {
+    // skip files that cannot be opened, due to permission issues etc
+    for entry in WalkDir::new(dir).into_iter().filter_map(|f| f.ok()) {
```
Comment (yahayaohinoyi) [inline src/main.rs]: why are we skipping files that can't be opened? Do we care about that? does the permission affect the deletion?

File: src/main.rs [hunk 5]
```diff
@@ -1,3 +1,89 @@
...
+        if path.is_file() {
+            let ext = path.extension()
+                .and_then(|ex| ex.to_str())
+                .unwrap_or("");
+
+            if paths_to_clear.iter().any(|p| ext == p) {
+                let metadata = fs::metadata(path)
+                    .with_context(|| format!("Failed to read metadata for file: {:?}", path))?;
+
+                if metadata.len() >= size {
```
Comment (yahayaohinoyi) [inline src/main.rs]: why are we doing this check?

File: src/main.rs [hunk 6]
```diff
@@ -1,3 +1,89 @@
...
+        let dir_str = temp_dir.path().to_str().unwrap().to_string();
+        directory_cleaner(&dir_str, &file_types, 500)?;
+
+        assert!(!file_path.exists(), "File should have been deleted");
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_directory_cleaner_should_not_delete_due_to_smaller_size() -> Result<()> {
```
Comment (yahayaohinoyi) [inline src/main.rs]: ditto

**Comment by yahayaohinoyi** [comment]
Love the structure; I think what we should extract the function `fn directory_cleaner(dir: &String, paths_to_clear: &[String], size: u64) ->` into a separate file/folder. For subsequent features, we can add in a new file that also houses its tests. This would help reduce conflict

**Comment by yahayaohinoyi** [comment]
Unrelated: Maybe we should change the CI a little bit to show the the test separately from the build. Gives more clarity. I'd send in a PR

**Comment by yahayaohinoyi** [comment]
What was the process of testing this manually?

File: src/arg.rs (hunk 1 above)
Comment (mubarak-ramadan) [reply src/arg.rs:1]: I'm not very familiar with rust's standard library but clap gives us more flexibility

File: src/arg.rs (hunk 2 above)
Comment (mubarak-ramadan) [reply src/arg.rs]: Resolved in latest

File: src/arg.rs (hunk 3 above)
Comment (mubarak-ramadan) [reply src/arg.rs:6]: Gives users flexibility to specifies the min file size to be deleted

File: src/main.rs (hunk 4 above)
Comment (mubarak-ramadan) [reply src/main.rs]: Now updated in latest

File: src/main.rs (hunk 5 above)
Comment (mubarak-ramadan) [reply src/main.rs]: To delete the files with size above the min_size provided by the user (defaults to O bytes if non is given)

Comment (yahayaohinoyi) [reply src/main.rs]: not part of the task ):

**Comment by yahayaohinoyi** [comment]
Why did you delete the cargo.lock file? I think this is useful no?

File: src/features/cleaner.rs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+            if paths_to_clear.iter().any(|p| ext == p) {
+                let metadata = fs::metadata(path)
+                    .with_context(|| format!("Failed to read metadata for file: {:?}", path))?;
+
+                if metadata.len() >= size {
+                    println!("Deleting file: {:?}", path);
+                    fs::remove_file(path)
+                        .with_context(|| format!("Failed to delete file: {:?}", path))?;
+                }
```
Comment (yahayaohinoyi) [inline src/features/cleaner.rs]:
```suggestion
                println!("Deleting file: {:?}", path);
                fs::remove_file(path)
                    .with_context(|| format!("Failed to delete file: {:?}", path))?;
```

File: src/main.rs [hunk 8]
```diff
@@ -1,3 +1,62 @@
-fn main() {
-    println!("Hello, World!");
+use anyhow::Result;
```
Comment (yahayaohinoyi) [inline src/main.rs:1]: Not too worried about this since it does the job. But It'd be better to use the built-in `Result<T, E>` enum

Comment (mubarak-ramadan) [reply src/main.rs:1]: This is neater given I can generically handle different type of error, but I get your point

**Review by yahayaohinoyi — APPROVED** [review]

**Pull request merged**